cassiopeia/preprocess/pipeline.py

```diff
@@ -34,6 +35,130 @@
 progress = tqdm
 
 
+class PreprocessError(Exception):
+    pass
+
+
+def convert_fastqs_to_unmapped_bam(
+    fastq_fps: List[str],
+    chemistry: Literal["dropseq", "10xv2", "10xv3", "slideseq2"],
+    output_directory: str,
+    name: Optional[str] = None,
+    n_threads: int = 1,
+) -> str:
+    """Converts FASTQs into an unmapped BAM based on a chemistry.
+
+    This function converts a set of FASTQs into an unmapped BAM with appropriate
+    BAM tags.
+
+    Args:
+        fastq_fps: List of paths to FASTQ files. Usually, this argument contains
+            two FASTQs, where the first contains the barcode and UMI sequences
+            and the second contains cDNA. The FASTQs may be gzipped.
+        chemistry: Sample-prep/sequencing chemistry used. The following
+            chemistries are supported:
+            * dropseq: Droplet-based scRNA-seq chemistry described in
+                Macosco et al. 2015.
+            * 10xv2: 10x Genomics 3' version 2
+            * 10xv3: 10x Genomics 3' version 3
+            * slideseq2: Slide-seq version 2
+        output_directory: The output directory where the unmapped BAM will be
+            written to. This directory must exist prior to calling this function.
+        name: Name of the reads in the FASTQs. This name is set as the read group
+            name for the reads in the output BAM, as well as the filename prefix
+            of the output BAM. If not provided, a short random UUID is used as
+            the read group name, but not as the filename prefix of the BAM.
+        n_threads: Number of threads to use. Defaults to 1.
+
+    Returns:
+        Path to written BAM
+
+    Raises:
+        PreprocessError if the provided chemistry does not exist.
+    """
+    if chemistry not in constants.CHEMISTRY_BAM_TAGS:
+        raise PreprocessError(f"Unknown chemistry {chemistry}")
+
+    logging.info("Converting FASTQs to unmapped BAM...")
+    t0 = time.time()
+
+    tag_map = constants.CHEMISTRY_BAM_TAGS[chemistry]
+    bam_fp = os.path.join(
+        output_directory, f"{name}_unmapped.bam" if name else "unmapped.bam"
+    )
+    ngs.fastq.fastqs_to_bam_with_chemistry(
+        fastq_fps,
+        ngs.chemistry.get_chemistry(chemistry),
+        tag_map,
+        bam_fp,
+        name=name,
+        show_progress=True,
+        n_threads=n_threads,
```

Comment:
Just making sure - `ngs` appropriately processes the FASTQs according to the chemistry, correct? Specifically, it knows how long cellBCs and UMIs are in the 10X v3 vs 10X v2 chemistry?

Reply:
Yes, it does. I also added specific tests for each chemistry.